E2E scale test for vSphere Cloud Provider Volume lifecycle operations

BaluDontu opened the pull request:
This PR adds an E2E test for vSphere Cloud Provider which will create/attach/detach/detach the volumes at scale with multiple threads based on user configurable values for number of volumes, volumes per pod and number of threads. (Since this is a scale test, number of threads would be low. This is only used to speed up the operation)

Test performs following tasks.

1. Create Storage Classes of 4 Categories (Default, SC with Non Default Datastore, SC with SPBM Policy, SC with VSAN Storage Capalibilies.)
2. Read VCP_SCALE_VOLUME_COUNT from System Environment.
3. Launch VCP_SCALE_INSTANCES go routines for creating VCP_SCALE_VOLUME_COUNT volumes. Each go routine is responsible for create/attach of VCP_SCALE_VOLUME_COUNT/VCP_SCALE_INSTANCES volumes.
4. Read VCP_SCALE_VOLUMES_PER_POD from System Environment. Each pod will be have VCP_SCALE_VOLUMES_PER_POD attached to it.
5. Once all the go routines are completed, we delete all the pods and volumes.

Which issue this PR fixes
fixes # vmware#291

```release-note
None
```

BaluDontu commented:
/assign @jeffvance

jeffvance commented on test/e2e/framework/pv_util.go:
```diff
@@ -790,12 +790,12 @@ func deletePD(pdName string) error {
 // Returns a pod definition based on the namespace. The pod references the PVC's
 // name.  A slice of BASH commands can be supplied as args to be run by the pod
-func MakePod(ns string, pvclaims []*v1.PersistentVolumeClaim, isPrivileged bool, command string) *v1.Pod {
+func MakePod(ns string, nodeSelector map[string]string, pvclaims []*v1.PersistentVolumeClaim, isPrivileged bool, command string) *v1.Pod {
```
nice addition

jeffvance commented:
In [vSphere pr](https://github.com/kubernetes/kubernetes/pull/53903#issuecomment-336557571) I made a suggestion about moving all of the vsphere tests into their own directory under _e2e/storage/_. Since we have another new vsphere test when will this move occur?

jeffvance commented on test/e2e/storage/vsphere_scale.go:
```diff
@@ -0,0 +1,242 @@
+var _ = SIGDescribe("vcp at scale [Feature:vsphere] ", func() {
+	f := framework.NewDefaultFramework("vcp-at-scale")
+
+	const (
```
These constants are duplicated in other vsphere tests. Can you consolidate>

BaluDontu replied:
Fixed!

jeffvance commented on test/e2e/storage/vsphere_scale.go:
```diff
@@ -0,0 +1,242 @@
+		namespace = f.Namespace.Name
+		nodeVolumeMapChan = make(chan map[string][]string)
+
+		volumeCountStr := os.Getenv("VCP_SCALE_VOLUME_COUNT")
```
Can you add comment listing all of the env variables needed to run these tests? Using env variables for specific tests seems to be unique to vsphere.

jeffvance replied:
Do you expect or care if these env variables are changed mid test or between `It` blocks? If the expectation is these env variable values are set once before running the vsphere tests then the `os.Getenv` calls do not need to be executed for each It. I realize right now there is only 1 It but there could be new tests added in the future. I think a better structure, assuming these env variables are static for the life of this test suite, is to assign the in the outer `SIGDescribe` block.

BaluDontu replied:
All missing env variables in comments section. Also moved os.Getenv to SIGDescribe block.

jeffvance commented on test/e2e/storage/vsphere_scale.go:
```diff
@@ -0,0 +1,242 @@
+
+		nodes = framework.GetReadySchedulableNodesOrDie(client)
+		if len(nodes.Items) < 2 {
+			framework.Skipf("Requires at least %d nodes (not %d)", 2, len(nodes.Items))
```
good err msg!

jeffvance commented on test/e2e/storage/vsphere_scale.go:
```diff
@@ -0,0 +1,242 @@
+				var scVSanParameters map[string]string
+				scVSanParameters = make(map[string]string)
+				scVSanParameters[Policy_HostFailuresToTolerate] = "1"
+				sc, err = client.StorageV1().StorageClasses().Create(getVSphereStorageClassSpec(storageclass2, scVSanParameters))
```
why can't the 4 calls to Create an sc be consolidated?

BaluDontu replied:
Fixed!

jeffvance commented on test/e2e/storage/vsphere_scale.go:
```diff
@@ -0,0 +1,242 @@
+			case storageclass1:
+				sc, err = client.StorageV1().StorageClasses().Create(getVSphereStorageClassSpec(storageclass1, nil))
+			case storageclass2:
+				var scVSanParameters map[string]string
```
This line and the following 3 can be generalized. The value name you're using here vs. in L130 has no special meaning. You can move this code outside of the switch and use a generic name.

BaluDontu replied:
Fixed!

jeffvance commented on test/e2e/storage/vsphere_scale.go:
```diff
@@ -0,0 +1,242 @@
+				scWithDatastoreSpec := getVSphereStorageClassSpec(storageclass4, scWithDSParameters)
+				sc, err = client.StorageV1().StorageClasses().Create(scWithDatastoreSpec)
+			}
+			Expect(sc).NotTo(BeNil())
```
can you add additional explanatory error texts to these two Expects?

BaluDontu replied:
Fixed!

jeffvance commented on test/e2e/storage/vsphere_scale.go:
```diff
@@ -0,0 +1,242 @@
+		framework.SkipUnlessProviderIs("vsphere")
+		client = f.ClientSet
+		namespace = f.Namespace.Name
+		nodeVolumeMapChan = make(chan map[string][]string)
```
where is this channel closed?

jeffvance commented on test/e2e/storage/vsphere_scale.go:
```diff
@@ -0,0 +1,242 @@
+
+		// Get the list of all volumes attached to each node from the go routines by reading the data from the channel
+		for instanceCount := 0; instanceCount < numberOfInstances; instanceCount++ {
+			for node, volumeList := range <-nodeVolumeMapChan {
```
doesn't this channel need to be explicitly closed to end the loop?

BaluDontu replied:
You needn't close every channel when you've finished with it. It's only necessary to close a channel when it is important to tell the receiving go routines that all data have been sent. A channel that the garbage collector determines to be unreachable will have its resources reclaimed whether or not it is closed.

However just to not end up consuming more resources, I have explicitly closed the channel.

dims commented:
/ok-to-test

BaluDontu commented:
@jeffvance I have addressed all your review comments and rebased the branch for conflicts. Please take a look at the latest changes.

Also regarding your request to move all the vsphere tests under a separate directory in e2e/storage/ will be taken as a separate PR after this change is merged. I do not want to mix moving files to separate directory as part of this scale test PR. I have an issue open for that - https://github.com/vmware/kubernetes/issues/353.

jeffvance commented on test/e2e/storage/vsphere_scale.go:
```diff
@@ -0,0 +1,230 @@
+		scArrays := make([]*storageV1.StorageClass, len(scNames))
+		for index, scname := range scNames {
+			// Create vSphere Storage Class
+			By(fmt.Sprintf("Creating Storage Class : %v", scname))
```
nit: I prefer %q for sc, pv, pvc, names...

BaluDontu replied:
Fixed!

jeffvance commented on test/e2e/storage/vsphere_scale.go:
```diff
@@ -0,0 +1,230 @@
+			case storageclass4:
+				scParams[Datastore] = datastoreName
+				sc, err = client.StorageV1().StorageClasses().Create(getVSphereStorageClassSpec(storageclass4, scParams))
+			}
```
In my previous comment about consolidating the sc Create call I was thinking more like:
```
for index, scname := range scNames {
    ,,,
    switch scname {
    case storageclass1:
        scParms = nil 
    case storageclass2:
        scParams[Policy_HostFailuresToTolerate] = "1"
    case storageclass3:
        scParams[SpbmStoragePolicy] = policyName
    case storageclass4:
        scParams[Datastore] = datastoreName
    }
sc, err = client.StorageV1().StorageClasses().Create(getVSphereStorageClassSpec(storageclass3, scParams))
...
```

BaluDontu replied:
Fixed!

jeffvance commented on test/e2e/storage/vsphere_scale.go:
```diff
@@ -0,0 +1,230 @@
+		vsp, err := vsphere.GetVSphere()
+		Expect(err).NotTo(HaveOccurred())
+
+		volumeCountPerInstance := volumeCount / numberOfInstances
```
where is the sanity checking on `numOfInstances`? It could be larger than `volumeCount`. Or it could be too large for the test env to create that many go routines.

BaluDontu replied:
Fixed!

jeffvance commented on test/e2e/storage/vsphere_scale.go:
```diff
@@ -0,0 +1,230 @@
+			}
+			Expect(sc).NotTo(BeNil(), "Storage class is empty")
+			Expect(err).NotTo(HaveOccurred(), "Failed to create storage class")
+			defer client.StorageV1().StorageClasses().Delete(scname, nil)
```
where is the `defer` for deleting the pods, pvcs, and pvs?

BaluDontu replied:
It is there in the it spec. As part of this test, we would want to delete all Pod's and PVC's at the end of test.

jeffvance replied:
There is no spec that I am aware of and not all tests do the cleanup the same. However, there are some helper functions (_framework/pv_util.go_ and _framework.util.go_) and example tests to copy.

BaluDontu replied:
This code deletes the pod and PVC's - https://github.com/kubernetes/kubernetes/pull/54018/files#diff-360422e42f5a2842e278d251d9d67839R152

jeffvance commented on test/e2e/upgrades/storage/persistent_volumes.go:
```diff
@@ -98,7 +98,7 @@ func (t *PersistentVolumeUpgradeTest) Teardown(f *framework.Framework) {
 
 // testPod creates a pod that consumes a pv and prints it out. The output is then verified.
 func (t *PersistentVolumeUpgradeTest) testPod(f *framework.Framework, cmd string) {
```
Doesn't this func already exist? Or maybe it is in another vpshere pr that I've reviewed?

BaluDontu replied:
I have not written this test as part of this PR. I have only modified it as there is a new param to Makepod().

jeffvance replied:
ack

BaluDontu commented:
@jeffvance : Can you take a look at latest commit that addresses your review comments.

jeffvance commented on test/e2e/storage/vsphere_scale.go:
```diff
@@ -0,0 +1,230 @@
+		for _, pod := range podList.Items {
+			pvcClaimList = append(pvcClaimList, getClaimsForPod(&pod, volumesPerPod)...)
+			By("Deleting pod")
+			framework.DeletePodWithWait(f, client, &pod)
```
err checking is missing

BaluDontu replied:
Fixed!

jeffvance commented:
Sorry, I missed that.

On Tue, Nov 7, 2017 at 2:40 PM, BaluDontu <notifications@github.com> wrote:

> *@BaluDontu* commented on this pull request.
> ------------------------------
>
> In test/e2e/storage/vsphere_scale.go
> <https://github.com/kubernetes/kubernetes/pull/54018#discussion_r149526309>
> :
>
> > +			switch scname {
> +			case storageclass1:
> +				sc, err = client.StorageV1().StorageClasses().Create(getVSphereStorageClassSpec(storageclass1, nil))
> +			case storageclass2:
> +				scParams[Policy_HostFailuresToTolerate] = "1"
> +				sc, err = client.StorageV1().StorageClasses().Create(getVSphereStorageClassSpec(storageclass2, scParams))
> +			case storageclass3:
> +				scParams[SpbmStoragePolicy] = policyName
> +				sc, err = client.StorageV1().StorageClasses().Create(getVSphereStorageClassSpec(storageclass3, scParams))
> +			case storageclass4:
> +				scParams[Datastore] = datastoreName
> +				sc, err = client.StorageV1().StorageClasses().Create(getVSphereStorageClassSpec(storageclass4, scParams))
> +			}
> +			Expect(sc).NotTo(BeNil(), "Storage class is empty")
> +			Expect(err).NotTo(HaveOccurred(), "Failed to create storage class")
> +			defer client.StorageV1().StorageClasses().Delete(scname, nil)
>
> This code deletes the pod and PVC's - https://github.com/kubernetes/
> kubernetes/pull/54018/files#diff-360422e42f5a2842e278d251d9d67839R152
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/54018#discussion_r149526309>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC7zIMyGT8TzGWdbpC3V4Ot4RdvKS9Ocks5s0NxQgaJpZM4P7VFU>
> .
>

BaluDontu commented:
I see the pull-kubernetes-unit failing not related to my change. Any thoughts?

BaluDontu committed:
vSphere scale tests

jeffvance commented:
maybe try running `hack/update-bazel.sh` and re-push?

On Tue, Nov 7, 2017 at 3:20 PM, k8s-ci-robot <notifications@github.com>
wrote:

> @BaluDontu <https://github.com/baludontu>: The following tests *failed*,
> say /retest to rerun them all:
> Test name Commit Details Rerun command
> pull-kubernetes-unit 3f813d9
> <https://github.com/kubernetes/kubernetes/commit/3f813d909a8748a972b59f61887ad47093958416>
> link
> <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/54018/pull-kubernetes-unit/63446/> /test
> pull-kubernetes-unit
> pull-kubernetes-verify 3f813d9
> <https://github.com/kubernetes/kubernetes/commit/3f813d909a8748a972b59f61887ad47093958416>
> link
> <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/54018/pull-kubernetes-verify/60644/> /test
> pull-kubernetes-verify
> pull-kubernetes-bazel-test 3f813d9
> <https://github.com/kubernetes/kubernetes/commit/3f813d909a8748a972b59f61887ad47093958416>
> link
> <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/54018/pull-kubernetes-bazel-test/14750/> /test
> pull-kubernetes-bazel-test
>
> Full PR test history <https://k8s-gubernator.appspot.com/pr/54018>. Your
> PR dashboard <https://k8s-gubernator.appspot.com/pr/BaluDontu>. Please
> help us cut down on flakes by linking to
> <https://github.com/kubernetes/community/blob/master/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests>
> an open issue
> <https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open> when
> you hit one in your PR.
>
> Instructions for interacting with me using PR comments are available here
> <https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md>.
> If you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://github.com/kubernetes/test-infra/blob/master/commands.md>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/54018#issuecomment-342656743>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC7zIAUN_UriOf4cJrPdQnvCsQIckygSks5s0OW_gaJpZM4P7VFU>
> .
>

BaluDontu commented:
/test pull-kubernetes-unit

BaluDontu commented:
@jeffvance : Can you provide your approval if everything looks good.

jeffvance commented on test/e2e/storage/vsphere_scale.go:
```diff
@@ -0,0 +1,247 @@
+		numberOfInstances, err = strconv.Atoi(numberOfInstancesStr)
+		Expect(err).NotTo(HaveOccurred(), "Error Parsing VCP_SCALE_INSTANCES")
+		Expect(numberOfInstances > 5).NotTo(BeTrue(), "Maximum allowed instances are 5")
+		Expect(numberOfInstances > volumeCount).NotTo(BeTrue(), "Number of instances should be less than the total volume count")
```
"less than or equal to" based on the condition.

jeffvance replied:
volumeCount is not validated to be within a required range, thus, this test for numberOfInstances is not reliable. In general, user input should be validated to prevent test flakes. It is ok with me if validating these env var inputs is addressed across all vshpere tests in a separate pr.

jeffvance commented:
I added another env var input comment that can be addressed in a separate pr.
Also, it is easier for reviewers if you push your review changes as a new commit so we can just focus on the new code. Then prior to a lgtm you can squash to a single commit.

/lgtm

mikedanese commented:
@kubernetes/sig-storage-pr-reviews

vishh commented:
/approve
